### DISCUSSION TITLE
Boolean operations on 2D sketches?

### DISCUSSION THREAD

**Discussion by paulftw** [sgenoud/replicad#43] [General]
I've checked the API docs but there seems to be no way to `intersect` rectangle and a circle for example.
This would be super useful when preparing a sketch for extrusion, adding NEMA mounting holes, etc.

What is the best way to implement this, apart from extruding each sketch and using 3D booleans?

I also really need [positive](https://en.wikibooks.org/wiki/OpenSCAD_User_Manual/Transformations#/media/File:OpenSCAD_offset_join-type_out.svg) & [negative](https://en.wikibooks.org/wiki/OpenSCAD_User_Manual/Transformations#/media/File:OpenSCAD_offset_join-type_in.svg) `offset` but that's probably a separate feature request.

**Comment by paulftw** [comment]
Answering my own question: `BlueprintSketcher` is pretty much a drop-in replacement of the basic sketcher, `intersectBlueprints` and `fuseBlueprints` can be used on its results.

Still trying to figure out an effective `offset` implementation

**Reply by raydeleu** [reply]
Did you check the new drawing API? If I remember well, the author of Replicad mentioned that the drawing API was created to allow boolean operations that are not possible with Sketches. See discussion on "drawing API". 

I edit this on my phone, I can give you more details later.

**Reply by sgenoud** [reply]
The `intersect` is missing for `Drawings`! I have created an issue to implement it.

The `Blueprint` is a lower level abstraction than the `Drawing` that does not easily take care of edge cases where there are discontinuities (or holes). I need to wrap the `intersectBlueprint` logic so that any two `Drawings` can be intersected together.

**Reply by paulftw** [reply]
thanks @raydeleu just found the drawing thread. It does look useful and approachable, will check it out for sure.

I was so happy when I discovered `Blueprint` and most of my experiments worked out of the box, now I wonder if I should start over with `Drawing` 😉
As a new user I'm finding it hard to grasp the difference between `Blueprint`, `Blueprints`, `CompoundBlueprint`, `(Compound?)Sketch(er?)`, `Drawing` and `DrawingPen`. but that's something that can be cleaned up later as the project matures.

**Reply by sgenoud** [reply]
The aim of the `Drawing` API is actually to cleanup the blueprint complexity. A drawing is either a `Blueprint` a `Blueprints` or a `CompoundBlueprint`. It then handles all the mess with this API (i.e. what happens when you try to fuse a `Blueprint` with a `CompoundBlueprint`.

All of this is hidden if you use `draw` do a bunch of edits on it and then you finish with a `done` (or `close`) and it creates a `Drawing` that you can do a bunch of things on.

Unless you are trying to do some lower level things, you don't need to worry about anything else.

Note that the Sketcher should disappear if we see that the `draw` API works well.

**Reply by paulftw** [reply]
@sgenoud you're doing a great job and I really appreciate your work!

not trying to spam you with features I might probably need some day, I'm sure you are already getting to many of those, but once intersect, union and difference are covered it'd be cool to consider inward/outward expansions and convex hulls - something I've personally needed in the past two weeks. 

Those functions are notable omissions from what openscad & jscad offer. Supporting them will make a transition to replicad easier.

(I did check how [cadquery is doing](https://github.com/CadQuery/cadquery/blob/2bf9116b5bba842248c113e4911b476f2dff5ea1/cadquery/hull.py#L367) convex hull and seems like they wrote the algorithm from scratch in python, so probably OCCT doesn't have a ready-to-use implementation).

There's also minkowski sum, but my non-scientific research into real-world openscad shows that regular people never heard of it.
